packages/table/src/interactions/selectable.tsx

```diff
@@ -256,71 +220,101 @@ export class DragSelectable extends React.Component<IDragSelectableProps, {}> {
+        const { onFocus } = this.props;
+        const focusedCellCoords = Regions.getFocusCellCoordinatesFromRegion(focusRegion);
+        onFocus(FocusedCellUtils.toFullCoordinates(focusedCellCoords, focusSelectionIndex));
     }
 
-    const lastRegion = regions[regions.length - 1];
-    const lastRegionCardinality = Regions.getRegionCardinality(lastRegion);
-    const regionCardinality = Regions.getRegionCardinality(region);
+    // Other
+    // =====
 
-    if (regionCardinality !== lastRegionCardinality) {
-        // TODO: add proper handling for expanding regions from one cardinality to another depending
-        // on the focused cell (see: https://github.com/palantir/blueprint/issues/823). for now,
-        // just return the new region by itself.
-        return [region];
+    private finishInteraction = () => {
+        BlueprintUtils.safeInvoke(this.props.onSelectionEnd, this.props.selectedRegions);
+        this.didExpandSelectionOnActivate = false;
     }
 
-    // simplified algorithm: expand the most recently selected region, and clear all others.
-    // TODO: pass the current focused cell into DragSelectable via props, then update this logic
-    // to leverage the focus cell's coordinates appropriately.
-    // (see: https://github.com/palantir/blueprint/issues/823)
-    if (regionCardinality === RegionCardinality.FULL_ROWS) {
-        const rowStart = Math.min(lastRegion.rows[0], region.rows[0]);
-        const rowEnd = Math.max(lastRegion.rows[1], region.rows[1]);
-        return [Regions.row(rowStart, rowEnd)];
-    } else if (regionCardinality === RegionCardinality.FULL_COLUMNS) {
-        const colStart = Math.min(lastRegion.cols[0], region.cols[0]);
-        const colEnd = Math.max(lastRegion.cols[1], region.cols[1]);
-        return [Regions.column(colStart, colEnd)];
-    } else if (regionCardinality === RegionCardinality.CELLS) {
-        const rowStart = Math.min(lastRegion.rows[0], region.rows[0]);
-        const colStart = Math.min(lastRegion.cols[0], region.cols[0]);
-        const rowEnd = Math.max(lastRegion.rows[1], region.rows[1]);
-        const colEnd = Math.max(lastRegion.cols[1], region.cols[1]);
-        return [Regions.cell(rowStart, colStart, rowEnd, colEnd)];
-    } else {
-        // if we've selected the FULL_TABLE, no need to expand it further.
-        return [region];
+    private expandSelectedRegions(regions: IRegion[], region: IRegion, focusedCell?: IFocusedCellCoordinates) {
+        if (regions.length === 0) {
+            return [region];
+        } else if (focusedCell != null) {
+            const expandedRegion = FocusedCellUtils.expandFocusedRegion(focusedCell, region);
+            return Regions.update(regions, expandedRegion);
+        } else {
+            const expandedRegion = Regions.expandRegion(regions[regions.length - 1], region);
+            return Regions.update(regions, expandedRegion);
+        }
```

Comment:
Pulled all the expand-region logic out into more easily unit-testable helper functions.